src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs

```diff
@@ -214,10 +214,20 @@ private SynthesizedBackingFieldSymbol CreateBackingField(bool isCreatedForFieldK
                 Interlocked.CompareExchange(ref _lazyBackingFieldSymbol, backingField, _lazyBackingFieldSymbolSentinel);
             }
 
-            return (SynthesizedBackingFieldSymbol)_lazyBackingFieldSymbol;
+            return (SynthesizedBackingFieldSymbol?)_lazyBackingFieldSymbol;
         }
 
-        internal SynthesizedBackingFieldSymbol CreateBackingFieldForFieldKeyword() => CreateBackingField(isCreatedForFieldKeyword: true, isEarlyConstructed: false);
+        internal SynthesizedBackingFieldSymbol? CreateBackingFieldForFieldKeyword(bool isSemanticModelBinder)
+        {
+            // field in the speculative model does not bind to a backing field if the original location was not a semi-auto property
+            var shouldBindField = !isSemanticModelBinder || FieldKeywordBackingField is not null;
```

Comment:
This does extra binding, which should be avoided if possible. I'm not sure how we can avoid it.

Reply:
This doesn't seem like the right condition to me. We don't care about whether we're in _any_ semantic model binder: if we're in a non-speculative context, then we want to bind the body if needed. It seems like this should be controlled by whether we're in a speculative model, and even then it could be that `field` should cause some binding to occur if the original property body has not yet been bound (this could happen when speculating without nullability enabled in a property body that has not yet been accessed in any way).

Reply:
More concretely, it seems like this should be done by the binding chain itself. If the current binder is a speculative model, then it should retrieve the next binder in the chain until it hits a binder that is not a speculative binder, and that binder should be the thing that causes or does not cause `field` to be created on the property symbol.

Reply:
@333fred Is there an existing way to know if the binder is speculative model? Or do I go ahead and add a new property?

Reply:
> it should retrieve the next binder in the chain until it hits a binder that is not a speculative binder, and that binder should be the thing that causes or does not cause `field` to be created on the property symbol.

Why not let `FieldKeywordBackingField` do the binding itself through its existing logic?

Reply:
>  if we're in a non-speculative context, then we want to bind the body if needed.

For non-speculative semantic models, the `FieldKeywordBackingField` check should cause binding, but I can see a potential cycle if this happened.

I tried to write a test to see what happens, but it seems I cannot get `IsSemanticModelBinder` to be true.

```csharp
        [Fact]
        public void TestFromSemanticModelBinder()
        {
            var comp = CreateCompilation(@"
public class C
{
    public int P1 { get => field; }
}
");
            var accessorBindingData = new SourcePropertySymbolBase.AccessorBindingData();
            comp.TestOnlyCompilationData = accessorBindingData;

            var tree = comp.SyntaxTrees[0];
            var model = comp.GetSemanticModel(tree);
            var info = model.GetSymbolInfo(tree.GetRoot().DescendantNodes().OfType<IdentifierNameSyntax>().Single());
            Assert.Empty(info.CandidateSymbols);
            Assert.False(info.IsEmpty);
            Assert.Equal("System.Int32 C.<P1>k__BackingField", info.Symbol.GetSymbol().ToTestDisplayString());
        }
```

Reply:
> Why not let FieldKeywordBackingField do the binding itself through its existing logic?

I'm not sure what you mean by this? How would `FieldKeywordBackingField` know whether to create a backing field or not? Basically, my intent is to make sure that https://github.com/dotnet/roslyn/blob/40ef62c7bae125af391bcc0bc3069f80e0692cad/src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs#L1564 is run by the appropriate binder: if the current binder is a speculative binder, then it needs to ask a non-speculative binder to bind the method body, and then asking the property for its field symbol should return `null` or the backing field, as appropriate.

> @333fred Is there an existing way to know if the binder is speculative model? Or do I go ahead and add a new property to Binder?

I don't believe there is an existing way, unless @AlekseyTs knows one that I don't?

Reply:
> How would FieldKeywordBackingField know whether to create a backing field or not?

@333fred

https://github.com/dotnet/roslyn/blob/1e155f44b27d228d313209f64bf8c3a578717905/src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs#L780-L784

Then `EnsureBackingFieldIsSynthesized` will do binding if necessary.

Reply:
Then perhaps the right solution is simply to adjust the line I mentioned in binding: if the current binder is a speculative binder, then it should call `FieldBackingKeyword`, which will trigger initial binding as appropriate. Otherwise, we want to call `CreateBackingFieldForFieldKeyword()`, so we avoid the cycle.

Reply:
@333fred I'd like us to have a test in place for what we think is broken. The one I tried to write earlier is currently passing. It'd be good to understand what's the broken scenario.

```csharp
            var comp = CreateCompilation(@"
public class C
{
    public int P1 { get => field; }
}
");
            var accessorBindingData = new SourcePropertySymbolBase.AccessorBindingData();
            comp.TestOnlyCompilationData = accessorBindingData;

            var tree = comp.SyntaxTrees[0];
            var model = comp.GetSemanticModel(tree);
            var info = model.GetSymbolInfo(tree.GetRoot().DescendantNodes().OfType<IdentifierNameSyntax>().Single());
            Assert.Empty(info.CandidateSymbols);
            Assert.False(info.IsEmpty);
            Assert.Equal("System.Int32 C.<P1>k__BackingField", info.Symbol.GetSymbol().ToTestDisplayString());
```

Things I observed:

- For the above test, the binder that binds "field" has a false `IsSemanticModelBinder`.
- `EnsureBackingFieldIsSynthesized` calls `getMethod.TryGetBodyBinder()` which I'm unsure whether it can return a semantic model binder. If it can't, then probably there is no problem with the current approach? (other than code refactoring moving the `shouldBindField` logic from symbol to binding). We'll end up with binding the body if needed anyway?

Specifically given there is currently no way of knowing whether a binder is speculative binder, I want to be sure we introduce a new property or flag for a good reason. Test coverage can also prevent future bugs and help developers understand why it was written this way.

Any suggestions on a test or corrections to my observations?